src/inject.cc

```diff
@@ -8,68 +8,119 @@
 #include <tlhelp32.h>
 #include <vector>
 
-bool InjectDll(const uint32_t ProcessId, const std::filesystem::path &Path) {
+
+HANDLE OpenRemoteProcess(const uint32_t ProcessId) {
   const uint32_t ProcessRights =
       PROCESS_CREATE_THREAD | PROCESS_QUERY_INFORMATION | PROCESS_VM_OPERATION |
       PROCESS_VM_WRITE | PROCESS_VM_READ;
   const HANDLE Process = OpenProcess(ProcessRights, false, ProcessId);
 
   if (Process == nullptr) {
-    return EXIT_FAILURE;
+    printf("Failed to open the remote process.\n");
+    return INVALID_HANDLE_VALUE;
   }
 
-  const PVOID RemoteDllPath = VirtualAllocEx(
-      Process, nullptr, 0x1000, MEM_COMMIT | MEM_RESERVE, PAGE_READWRITE);
+  return Process;
+}
+
+PVOID AllocateMemInRemoteProcess(HANDLE hRemoteProcess,
+                                 const std::filesystem::path &Path) {
+  const std::string DllPath = Path.string();
+  const size_t DllPathLen = DllPath.size() + 1;
+  SIZE_T BytesWritten;
+
+  const PVOID RemoteDllPath = VirtualAllocEx(hRemoteProcess, nullptr, 0x1000, MEM_COMMIT | MEM_RESERVE,
+                     PAGE_READWRITE);
 
   if (RemoteDllPath == nullptr) {
     printf("VirtualAllocEx failed.\n");
-    return EXIT_FAILURE;
+    goto Error;
   }
 
-  const std::string DllPath = Path.string();
-  const size_t DllPathLen = DllPath.size() + 1;
-  SIZE_T BytesWritten;
-  if (!WriteProcessMemory(Process, RemoteDllPath, DllPath.c_str(), DllPathLen,
+  if (!WriteProcessMemory(hRemoteProcess, RemoteDllPath, DllPath.c_str(),
+                          DllPathLen,
                           &BytesWritten)) {
+    VirtualFreeEx(hRemoteProcess, RemoteDllPath, 0, MEM_RELEASE);
     printf("WriteProcessMemory failed.\n");
-    return EXIT_FAILURE;
+    goto Error;
   }
 
+  return RemoteDllPath;
+
+Error:
+  CloseHandle(hRemoteProcess);
+  return nullptr;
+}
+
+HANDLE CreateThreadInRemoteProcess(HANDLE hRemoteProcess, PVOID RemoteDllPath) {
+  PVOID LoadLibraryA;
+  HANDLE Thread;
+  DWORD Tid;
+
   const HMODULE Kernelbase = GetModuleHandleA("kernelbase");
   if (Kernelbase == nullptr) {
     printf("GetModuleHandleA failed.\n");
-    return EXIT_FAILURE;
+    goto Error;
   }
 
-  const PVOID LoadLibraryA = PVOID(GetProcAddress(Kernelbase, "LoadLibraryA"));
+  LoadLibraryA = PVOID(GetProcAddress(Kernelbase, "LoadLibraryA"));
```

Comment:
nit: Can you either directly move the definition + declaration there? `PVOID LoadLibraryA = PVOID(...` maybe even `const` it if it is ok?

Reply:
And same for Thread / Tid.

Reply:
I am not a huge fan of all declaration at the top of the block :-D